[Neo Jing Xuan] iP

neojxuan opened the pull request:
# Duke

> “The trouble with having an open mind, of course, is that people will insist on coming along and trying to put things in it.”
> ― Terry Pratchett, Diggers

Duke will not try to put things into your mind. Instead, these are the things *you* can put into Duke:
- Todos :heavy_check_mark:
- Events :ticket:
- Deadlines :scream: 
- ~~Others~~

All you need to do is,

1. download it from [here](https://github.com/neojxuan/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :sleeping:

And it is **FREE!**

Features:

- [x] Adding and deleting tasks
- [x] Mark tasks as done
- [x] Searching tasks by keywords
- [ ] Editing tasks

For programmers, this is the `main` method that launches the application. It can be found in the `Launcher` class:
```java
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```

Commit:
Add Gradle support

neojxuan committed:
add greet, echo and exit features

neojxuan committed:
add ability to store and list user input

neojxuan committed:
add mark tasks as done feature

neojxuan committed:
add different types of tasks

neojxuan committed:
fix task type and printed messages

neojxuan committed:
add automated text ui testing

neojxuan committed:
add error handling for incorrect user inputs

neojxuan committed:
add delete task function and refactor to use arraylist

neojxuan committed:
add command enum to store valid user commands

neojxuan committed:
refactor Command enum: store name

neojxuan committed:
add tests for invalid input and delete

neojxuan committed:
Add saving and loading tasks from hard disc

neojxuan committed:
Add DateTime in event and deadline tasks

neojxuan committed:
Merge branch 'branch-Level-7'

neojxuan committed:
Merge branch 'branch-Level-8'

neojxuan committed:
refactor: abstract ui, storage, parser, tasklist classes

neojxuan committed:
organise classes into packages

neojxuan committed:
tests: add JUnit tests for Parser and TaskList

neojxuan committed:
feat: add find tasks by keyword feature

neojxuan committed:
docs: Add JavaDocs to classes and public methods

neojxuan committed:
style: Fix style to comply with Coding Standard.

neojxuan committed:
docs: Add JavaDocs for classes and public methods

neojxuan committed:
style: Fix style to follow coding standard

neojxuan committed:
feat: Add search for tasks in list by keywords

marcuspeh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+    
+    public static void main(String[] args) {
```
Perhaps you can add in a javadocs comment here since all methods other than getter and setters need to have it?

marcuspeh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+        if (answer.contains(" ")) {
+            taskDetails = answer.substring(answer.indexOf(" ") + 1);
+        }
+        switch (command) {
```
I think you can remove the indentation for case and default of switch?
![image](https://user-images.githubusercontent.com/54053213/131605144-88758d3d-4014-4133-b14d-7f88abb7a752.png)

marcuspeh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+    private static Todo parseTodo(String taskDetails) throws DukeException {
+        checkEmptyTaskDetails(taskDetails);
+        return new Todo(taskDetails);
+    }
+
+    private static Event parseEvent(String taskDetails) throws DukeException {
+        checkEmptyTaskDetails(taskDetails);
+        String[] parts = taskDetails.split(" /at ");
+        if (parts.length < 2) {
+            throw new DukeException("Event descriptions must contain /at [dd-mm-yyyy hh:mm]");
+        }
+        String description = parts[0];
+        String at = parts[1];
+        return new Event(description, at);
+    }
+
+    private static Deadline parseDeadline(String taskDetails) throws DukeException {
+        checkEmptyTaskDetails(taskDetails);
+        String[] parts = taskDetails.split(" /by ");
+        if (parts.length < 2) {
+            throw new DukeException("Deadline descriptions must contain /by [dd-mm-yyyy hh:mm]");
+        }
+        String description = parts[0];
+        String by = parts[1];
+        return new Deadline(description, by);
+    }
+
+    protected static int getTaskIndex(String answer) throws DukeException {
+        String taskNo = answer.substring(answer.indexOf(" ") + 1);
+        try {
+            int taskIndex = Integer.parseInt(taskNo) - 1;
+            if (taskIndex < 0) {
+                throw new DukeException("Invalid task number. Task number should be positive.");
+            }
+            return taskIndex;
+        } catch (NumberFormatException e) {
+            throw new DukeException("Invalid task number. Sample input with correct format: [command] [taskNo]"
+                    + " eg. 'done 2'");
+        }
+    }
```
Maybe you should include a comment for each method?

marcuspeh commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,111 @@
+    private final String filePath;
+
+    /**
+     * Constructs a Storage that saves and loads tasks from the file specified by filePath/
```
Is there any reason why you put a / at the back of the comment?

marcuspeh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,141 @@
+        return taskStrings;
+    }
+    
+    public TaskList findTasksWithKeyword(String keyword) {
```
Perhaps you can include a comment here?

marcuspeh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,141 @@
+        return new TaskList(arrListTasks);
+    }
+    
+    private void validateTaskIndex(int taskIndex) throws DukeException { 
```
Same as above, I think a comment will be useful here.

marcuspeh reviewed (commented):
Generally followed the java convention quite well. Just need to take note on comments for classes and you should be fine.

neojxuan committed:
update gitignore and add minifest file for version 0.1

neojxuan committed:
Merge branch 'add-gradle-support'

neojxuan committed:
update gradle build file's application mainClassName

neojxuan committed:
style: add checkstyle config and fix style errors

nicole-luo-exe commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,113 @@
+package duke;
```
Well abstracted methods make for easy understanding. Exceptions are also properly handled. 💯

nicole-luo-exe commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+        ui.showWelcome();
+        boolean isExit = false;
+
+        while (!isExit) {
```
Perhaps could consider `while (!c.isExit) { `?

nicole-luo-exe commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    private static Task parseTask(String fileLine, int lineNo) throws DukeException {
```
Perhaps the responsibility of parsing might be more suited for the Parser class?

nicole-luo-exe reviewed (commented):
Code is of quality. 😺 
SLAP observed, nesting is kept minimal, methods kept short (within 30 LoC).
I really liked the generous use of abstractions. It made reading through your code very pleasant! Keep up the good work! 👍

neojxuan committed:
Add JavaFX Part 1

neojxuan committed:
Add JavaFX Part 2

neojxuan committed:
Add JavaFX Part 3

neojxuan committed:
add javafx level 4 gui for duke

neojxuan committed:
Add JavaFX GUI

neojxuan committed:
Add assertions

Project code does not contain assertions. There may be incorrect
assumptions made in the code.

Assertions can be used to verify assumptions made in the programme.
They help to make sure that the programme behaves as expected.

Let's add assertions to:
* ensure that parsed user input and duke's response are never null
* storage file is correctly created and can be read
* private method in TaskList validates task index correctly

neojxuan committed:
Improve code quality

Methods in Parser and Storage classes have different levels of
abstraction.

Having a single level of abstraction can make code neater and
improve readability.

Let's reorganise the code in the Parser and Storage classes by
extracting methods to follow SLAP.

neojxuan committed:
throw exceptions if datetime cannot be parsed for deadline and event tasks

neojxuan committed:
Improve code quality

Extract methods to follow SLAP, update Javadocs and throw exceptions where appropriate.

neojxuan committed:
Merge remote-tracking branch 'origin/master' into branch-A-Assertions

neojxuan committed:
Merge add assertions  branch

neojxuan committed:
Add edit command to enable user to edit description and date of tasks

neojxuan committed:
Merge branch 'branch-C-Update'

neojxuan committed:
Add javadocs for public methods

neojxuan committed:
Fix checkstyle errors

neojxuan committed:
Improve gui, add background colours and padding

neojxuan committed:
Show welcome message when application is launched

neojxuan committed:
Update profile pictures for Duke and User in gui

neojxuan committed:
Add Ui image for website

neojxuan committed:
Fix: Close gui window when user enters bye

neojxuan committed:
add comments in parseEditCommand function

neojxuan committed:
Save tasks to storage file after every command that changes tasks

neojxuan committed:
Add user guide

neojxuan committed:
Add contents sections with links to user guide

neojxuan committed:
Add datetime input format to user guide

neojxuan committed:
Add restrictions on task index in user guide

neojxuan committed:
Set window title for gui

neojxuan committed:
Update Ui.png with Application Title

neojxuan committed:
Refactor parser to have clearer parameter names and error messages, extract methods

neojxuan committed:
Add exception if user tries to mark already done task as done again

neojxuan committed:
Change test names to follow coding standard

neojxuan committed:
Fix errors in expected outcomes in user guide

neojxuan committed:
Fix out of bounds exception thrown by edit command

neojxuan committed:
Fix checkstyle errors

neojxuan committed:
Fix javadocs and minor refactoring

neojxuan committed:
Update README to include Quick Start section
